File: 00-每日一题/20200408_290.py
# 查找表
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        n_patt = len(pattern)
        s_list = str.split(' ')
        n_s = len(s_list)
        if n_patt != n_s:
            return False
        ps_map = dict()
        pattern_set = set()
        for i in range(n_patt):
            if s_list[i] not in ps_map:
                if pattern[i] in pattern_set:
                    return False  # 使用了相同的pattern
                ps_map[s_list[i]] = pattern[i]
                pattern_set.add(pattern[i])
            else:
                if ps_map[s_list[i]] != pattern[i]:
                    return False
        return True


if __name__ == '__main__':
    obj = Solution()
    while True:
        pattern = input().strip()
        s = input().strip()
        res = obj.wordPattern(pattern, s)
        print(res)

File: 00-每日一题/20200408_290.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char,string> cmap;
        unordered_map<string,char> smap;
        int i=0;
        int j=0;
        while(i<pattern.size() && j<str.size())
        {
            char a = pattern[i];
            string temp ="";
            while(str[j]!=' ' && j<str.size())
            {
                temp+=str[j];
                j++; 
            }
            j++;
            if(cmap.count(a)==0 && smap.count(temp)==0)
            {
                cmap[a] = temp;
                smap[temp] = a;
            }
            else if(cmap[a]!=temp)
                return false;
            i++;
        }
        if(i<pattern.size() || j<str.size())
            return false;
        return true;
       
    }
};